programs/mango-v4/src/instructions/serum3_place_order.rs

```diff
@@ -233,24 +167,65 @@ pub fn serum3_place_order(
     //
     // After-order tracking
     //
+    let after_oo_free_slots;
     let after_oo = {
         let oo_ai = &ctx.accounts.open_orders.as_ref();
         let open_orders = load_open_orders_ref(oo_ai)?;
+        after_oo_free_slots = open_orders.free_slot_bits;
         OpenOrdersSlim::from_oo(&open_orders)
     };
     let oo_difference = OODifference::new(&before_oo, &after_oo);
 
+    //
+    // Track the highest bid and lowest ask, to be able to evaluate worst-case health even
+    // when they cross the oracle
+    //
+    let serum = account.serum3_orders_mut(serum_market.market_index)?;
+    if !before_had_bids {
+        // The 0 state means uninitialized/no value
+        serum.highest_placed_bid_inv = 0.0;
+    }
+    if !before_had_asks {
+        serum.lowest_placed_ask = 0.0;
+    }
+    let new_order_on_book = after_oo_free_slots != before_oo_free_slots;
+    if new_order_on_book {
+        match side {
+            Serum3Side::Ask => {
+                // in the normal quote per base units
+                let limit_price =
+                    limit_price_lots as f64 * quote_lot_size as f64 / base_lot_size as f64;
+                serum.lowest_placed_ask = if serum.lowest_placed_ask == 0.0 {
```

Comment:
why is 0 the reserved value instead of something impossible to input: e.g. NaN?

Reply:
what is the impact of placing one big ask at $0 and a small one at $100?

Reply:
Zero is the reserved value for migration reasons: All existing Serum3Orders structs will have 0.0 in these fields, so we need to special case for it anyway.

You can't place orders at 0. But if you placed an ask at $0.001 or something and another at $100 then both reserved base amounts would be valued at $0.001. This is an intentional underestimation we make because we can't track the true prices and quantities.

Reply:
gotcha